#! /bin/sh /usr/share/dpatch/dpatch-run
## 30_olympus_makernote.dpatch by  <fpeters@debian.org>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Added support for Olympus S760 & S770 makernote (upstream bug #1703284)
## DP: (closes: #418945)

@DPATCH@
diff -urNad libexif-0.6.13~/libexif/exif-data.c libexif-0.6.13/libexif/exif-data.c
--- libexif-0.6.13~/libexif/exif-data.c	2007-05-08 12:00:57.000000000 +0200
+++ libexif-0.6.13/libexif/exif-data.c	2007-05-08 12:05:15.156112051 +0200
@@ -665,8 +665,8 @@
 		return EXIF_DATA_TYPE_MAKER_NOTE_NONE;
 
 	/* Olympus & Nikon */
-	if ((e->size >= 5) && (!memcmp (e->data, "OLYMP", 5) ||
-			       !memcmp (e->data, "Nikon", 5)))
+	if ((e->size >= 8) && (!memcmp (e->data, "OLYMP", 6) ||
+			!memcmp (e->data, "OLYMPUS", 8) || !memcmp (e->data, "Nikon", 6)))
 		return EXIF_DATA_TYPE_MAKER_NOTE_OLYMPUS;
 
 	em = exif_data_get_entry (d, EXIF_TAG_MAKE);
diff -urNad libexif-0.6.13~/libexif/olympus/exif-mnote-data-olympus.c libexif-0.6.13/libexif/olympus/exif-mnote-data-olympus.c
--- libexif-0.6.13~/libexif/olympus/exif-mnote-data-olympus.c	2005-07-11 22:20:29.000000000 +0200
+++ libexif-0.6.13/libexif/olympus/exif-mnote-data-olympus.c	2007-05-08 12:05:57.658534126 +0200
@@ -86,7 +86,7 @@
 		unsigned char **buf, unsigned int *buf_size)
 {
 	ExifMnoteDataOlympus *n = (ExifMnoteDataOlympus *) ne;
-	unsigned int i, o, s, doff, base = 0, o2 = 6;
+	unsigned int i, o, s, doff, base = 0, o2 = 6 + 2;
 	int datao = 0;
 
 	if (!n || !buf || !buf_size) return;
@@ -95,9 +95,8 @@
 	 * Allocate enough memory for all entries and the number of entries.
 	 */
 	*buf_size = 6 + 2 + 2 + n->count * 12;
-	o2 += 2;
 	switch (n->version) {
-	case olympus: 
+	case olympusV1:
 		*buf = exif_mem_alloc (ne->mem, *buf_size);
 		if (!*buf) return;
 
@@ -105,6 +104,20 @@
 		strcpy ((char *)*buf, "OLYMP");
 		datao = n->offset;
 		break;
+	case olympusV2:
+		*buf_size += 8-6 + 4;
+		*buf = exif_mem_alloc (ne->mem, *buf_size);
+		if (!*buf) return;
+
+		/* Write the header and the number of entries. */
+		strcpy ((char *)*buf, "OLYMPUS");
+		exif_set_short (*buf + 8, n->order, (ExifShort) (
+			(n->order == EXIF_BYTE_ORDER_INTEL) ?
+			('I' << 8) | 'I' :
+			('M' << 8) | 'M'));
+		exif_set_short (*buf + 10, n->order, (ExifShort) 3);
+		o2 += 4;
+		break;
 	case nikonV1: 
 		base = MNOTE_NIKON1_TAG_BASE;
 
@@ -201,14 +214,31 @@
 	 * lastly 0x2A.
 	 */
 	if (buf_size - n->offset < 22) return;
-	if (!memcmp (buf + o2, "OLYMP", 5)) {
+	if (!memcmp (buf + o2, "OLYMP", 6)) {
 		exif_log (en->log, EXIF_LOG_CODE_DEBUG, "ExifMnoteDataOlympus",
-			"Parsing Olympus maker note...");
+			"Parsing Olympus maker note v1...");
 
 		/* The number of entries is at position 8. */
-		n->version = olympus;
+		n->version = olympusV1;
 		o2 += 8;
 
+	} else if (!memcmp (buf + o2, "OLYMPUS", 8)) {
+		/* Olympus S760, S770 */
+		datao = o2;
+		o2 += 8;
+		exif_log (en->log, EXIF_LOG_CODE_DEBUG, "ExifMnoteDataOlympus",
+			"Parsing Olympus maker note v2 (0x%02x, %02x, %02x, %02x)...",
+			buf[o2], buf[o2 + 1], buf[o2 + 2], buf[o2 + 3]);
+
+		if ((buf[o2] == 'I') && (buf[o2 + 1] == 'I'))
+			n->order = EXIF_BYTE_ORDER_INTEL;
+		else if ((buf[o2] == 'M') && (buf[o2 + 1] == 'M'))
+			n->order = EXIF_BYTE_ORDER_MOTOROLA;
+
+		/* The number of entries is at position 8+4. */
+		n->version = olympusV2;
+		o2 += 4;
+
 	} else if (!memcmp (buf + o2, "Nikon", 6)) {
 		o2 += 6;
 		exif_log (en->log, EXIF_LOG_CODE_DEBUG, "ExifMnoteDataOlympus",
diff -urNad libexif-0.6.13~/libexif/olympus/exif-mnote-data-olympus.h libexif-0.6.13/libexif/olympus/exif-mnote-data-olympus.h
--- libexif-0.6.13~/libexif/olympus/exif-mnote-data-olympus.h	2005-07-11 22:20:29.000000000 +0200
+++ libexif-0.6.13/libexif/olympus/exif-mnote-data-olympus.h	2007-05-08 12:05:15.156112051 +0200
@@ -26,7 +26,7 @@
 #include <libexif/exif-byte-order.h>
 #include <libexif/exif-mem.h>
 
-enum OlympusVersion {olympus = 0, nikonV1 = 1, nikonV2 = 2};
+enum OlympusVersion {nikonV1 = 1, nikonV2 = 2, olympusV1 = 3, olympusV2 = 4 };
 
 
 typedef struct _ExifMnoteDataOlympus ExifMnoteDataOlympus;
diff -urNad libexif-0.6.13~/libexif/olympus/mnote-olympus-entry.c libexif-0.6.13/libexif/olympus/mnote-olympus-entry.c
--- libexif-0.6.13~/libexif/olympus/mnote-olympus-entry.c	2005-08-15 22:24:05.000000000 +0200
+++ libexif-0.6.13/libexif/olympus/mnote-olympus-entry.c	2007-05-08 12:05:15.156112051 +0200
@@ -432,7 +432,7 @@
 		break;
 	case MNOTE_OLYMPUS_TAG_INFO:
 		CF (entry->format, EXIF_FORMAT_ASCII, v, maxlen);
-		CC2 (entry->components, 52, 53, v, maxlen);
+		CC2 (entry->components, 52, 60, v, maxlen);
 		strncpy (v, (char *)entry->data, MIN (maxlen, entry->size));
 		break;
 	case MNOTE_OLYMPUS_TAG_ID:
